overlord/snapstate/snapmgr.go

```diff
@@ -100,8 +113,53 @@ func (m *SnapManager) doInstallSnap(t *state.Task, _ *tomb.Tomb) error {
 	t.State().Unlock()
 
 	pb := &TaskProgressAdapter{task: t}
-	_, err := m.backend.Install(inst.Name, inst.Channel, inst.Flags, pb)
-	return err
+	downloadedSnapFile, developer, err := m.backend.Download(inst.Name, inst.Channel, pb)
+	if err != nil {
+		return err
+	}
+	dl.DownloadedSnap = downloadedSnapFile
+	dl.Developer = developer
+
+	// update instState for the next task
+	t.State().Lock()
+	t.Set("download-state", dl)
+	t.State().Unlock()
+
+	return nil
+}
+
+func (m *SnapManager) doInstallLocalSnap(t *state.Task, _ *tomb.Tomb) error {
+	var inst installState
+	var dl downloadState
+	var dlTaskID string
+
+	t.State().Lock()
+	if err := t.Get("install-state", &inst); err != nil {
+		return err
+	}
+	t.State().Unlock()
+
+	// local snaps are special
+	if inst.LocalFileName != "" {
```

Comment:
The two recommended changes above make an improvement here more obvious:
```
var snapPath string
var developer string
if inst.SnapPath != "" {
        snapPath = inst.SnapPath
} else if inst.DownladTaskId != "" {
        ...
        ...
} else {
        return fmt.Errorf("internal error: install-snap created without a snap path source")
}
...
pb := &TaskProgressAdapter{task: t}
return m.backend.InstallLocal(snapPath, developer, inst.Flags, pb)
```

In other words.. the two paths are very much alike rather than special. It's just a different choice for where to pick the snap path from.